lib/pmdtester/report_diff.rb

```diff
@@ -1,185 +1,210 @@
+      @violation_diffs_by_rule = {}
 
-      @violation_diffs = build_diffs(base_violations.violations, patch_violations.violations)
-      @violation_diffs = merge_changed_violations(@violation_diffs)
+      @base_report = base_report
+      @patch_report = patch_report
 
-      @new_violations_size,
-          @changed_violations_size,
-          @removed_violations_size = get_diffs_size(@violation_diffs)
-      @violation_diffs_size = @new_violations_size +
-                              @changed_violations_size +
-                              @removed_violations_size
+      @rule_infos_union = base_report.infos_by_rule.dup
+      @violation_diffs_by_file = {}
+      @error_diffs_by_file = {}
+
+      diff_with(patch_report)
     end
 
-    def calculate_errors(base_errors, patch_errors)
-      @base_errors_size = base_errors.errors_size
-      @patch_errors_size = patch_errors.errors_size
-      @error_diffs = build_diffs(base_errors.errors, patch_errors.errors)
-      @new_errors_size, _, @removed_errors_size = get_diffs_size(@error_diffs)
-      @error_diffs_size = @new_errors_size + @removed_errors_size
+    def rule_summaries
+      @violation_diffs_by_rule.map do |(rule, counters)|
+        {
+          'name' => rule,
+          'info_url' => @rule_infos_union[rule].info_url,
+          **counters.to_h
+        }
+      end
     end
 
-    def calculate_configerrors(base_configerrors, patch_configerrors)
-      @base_configerrors_size = base_configerrors.size
-      @patch_configerrors_size = patch_configerrors.size
-      @configerrors_diffs = build_diffs(base_configerrors.errors, patch_configerrors.errors)
-      @new_configerrors_size, _, @removed_configerrors_size = get_diffs_size(@configerrors_diffs)
-      @configerrors_diffs_size = @new_configerrors_size + @removed_configerrors_size
+    private
+
+    def diff_with(patch_report)
+      @violation_counts.patch_total = patch_report.violations_by_file.total_size
+      @error_counts.patch_total = patch_report.errors_by_file.total_size
+
+      @violation_diffs_by_file = build_diffs(@violation_counts, &:violations_by_file)
+      count(@violation_diffs_by_file) { |v| getvdiff(v.rule_name) } # record the diffs in the rule counter
+
+      @error_diffs_by_file = build_diffs(@error_counts, &:errors_by_file)
+
+      count_by_rule(@base_report.violations_by_file, base: true)
+      count_by_rule(@patch_report.violations_by_file, base: false)
+      self
     end
 
-    def calculate_details(base_info, patch_info)
-      base_details = PmdReportDetail.new
-      base_details.load(base_info) unless base_info.nil?
-      patch_details = PmdReportDetail.new
-      patch_details.load(patch_info) unless patch_info.nil?
+    def record_rule_info(violation)
+      return if @rule_infos_union.key?(violation.rule_name)
+
+      @rule_infos_union[violation.rule_name] = RuleInfo.new(violation.rule_name, violation.info_url)
```

Comment:
True, but I don't think there is language information in the XML report format. We could instead compare the URL I guess (but it might be missing..). This is maybe something to keep in mind for pmd/pmd#2693 but not to do right now?

Reply:
Yes, we might need to change something, not right now.